### PR TITLE
Delete snapshot ongoing clone split

### PR DESCRIPTION
**Pull request by kpawar-sap**

### CODE REVIEW

File: manila/share/drivers/netapp/dataontap/client/client_cmode.py [hunk 1]
```diff
@@ -4095,17 +4095,22 @@ def delete_snapshot(self, volume_name, snapshot_name, ignore_owners=False):
...
     @na_utils.trace
-    def soft_delete_snapshot(self, volume_name, snapshot_name):
+    def soft_delete_snapshot(self, volume_name, snapshot_name,
+                             vserver_client, snapshot_children=[]):
         """Deletes a volume snapshot, or renames it if delete fails."""
         try:
             self.delete_snapshot(volume_name, snapshot_name)
         except netapp_api.NaApiError:
+            for snapshot_child in snapshot_children:
+                vserver_client.volume_clone_split_stop(snapshot_child['name'])
```
[the remark this answers is not in the document]
Comment (kpawar-sap) [reply manila/share/drivers/netapp/dataontap/client/client_cmode.py:4105]: I was under impression split start needed for both conditions. But if its not needed in 'if is_flexgroup' we can move down further and optimize it.

[the remark this answers is not in the document]
Comment (kpawar-sap) [reply manila/share/drivers/netapp/dataontap/client/client_cmode.py]: changed and pushed

**Review by kpawar-sap — COMMENTED** [review]
please review

File: manila/share/drivers/netapp/dataontap/client/client_cmode.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Carthaca) [reply manila/share/drivers/netapp/dataontap/client/client_cmode.py]: I'm good with the 'else' condition optimisation now,
but it is still needed in the 'if is_flexgroup' condition.

[the remark this answers is not in the document]
Comment (kpawar-sap) [reply manila/share/drivers/netapp/dataontap/client/client_cmode.py]: fixed

File: manila/share/drivers/netapp/dataontap/client/client_cmode.py [hunk 2]
```diff
@@ -4107,6 +4107,12 @@ def soft_delete_snapshot(self, volume_name, snapshot_name):
             msg_args = {'snapshot': snapshot_name, 'volume': volume_name}
             LOG.info(msg, msg_args)
 
+            # Snapshots are locked by clone(s), so split the clone(s)
+            snapshot_children = vserver_client.get_clone_children_for_snapshot(
```
Comment (Carthaca) [inline manila/share/drivers/netapp/dataontap/client/client_cmode.py:4111]:
```suggestion
            snapshot_children = self.get_clone_children_for_snapshot(
```

File: manila/share/drivers/netapp/dataontap/client/client_cmode.py [hunk 3]
```diff
@@ -4107,6 +4107,12 @@ def soft_delete_snapshot(self, volume_name, snapshot_name):
...
+                volume_name, snapshot_name)
+            for snapshot_child in snapshot_children:
+                vserver_client.volume_clone_split_start(snapshot_child['name'])
```
Comment (Carthaca) [inline manila/share/drivers/netapp/dataontap/client/client_cmode.py:4114]:
```suggestion
                self.volume_clone_split_start(snapshot_child['name'])
```

File: manila/share/drivers/netapp/dataontap/client/client_cmode.py [hunk 4]
```diff
@@ -4095,7 +4095,7 @@ def delete_snapshot(self, volume_name, snapshot_name, ignore_owners=False):
         self.send_request('snapshot-delete', api_args)
 
     @na_utils.trace
-    def soft_delete_snapshot(self, volume_name, snapshot_name):
+    def soft_delete_snapshot(self, volume_name, snapshot_name, vserver_client):
```
Comment (Carthaca) [inline manila/share/drivers/netapp/dataontap/client/client_cmode.py:4098]:
```suggestion
    def soft_delete_snapshot(self, volume_name, snapshot_name):
```

File: manila/share/drivers/netapp/dataontap/cluster_mode/lib_base.py [hunk 5]
```diff
@@ -2018,21 +2018,22 @@ def _delete_snapshot(self, vserver_client, share_name, snapshot_name,
...
                 # NOTE(felipe_rodrigues): ONTAP does not allow rename a
                 # FlexGroup snapshot, so it cannot be soft deleted. It will
                 # wait for all split clones complete.
                 self._delete_busy_snapshot(vserver_client, share_name,
                                            snapshot_name)
             else:
-                vserver_client.soft_delete_snapshot(share_name, snapshot_name)
-
+                vserver_client.soft_delete_snapshot(
+                    share_name, snapshot_name, vserver_client)
```
Comment (Carthaca) [inline manila/share/drivers/netapp/dataontap/cluster_mode/lib_base.py:2036]:
```suggestion
                    share_name, snapshot_name)
```

**Review by Carthaca — CHANGES_REQUESTED** [review]

**Review by Carthaca — APPROVED** [review]

**Pull request merged**